src/ets/scala/io/iohk/ethereum/ets/blockchain/ScenarioSetup.scala

```diff
@@ -0,0 +1,129 @@
+          "decoded genesis block header did not match the expectation")
+        block
+
+      case None =>
+        Block(scenario.genesisBlockHeader.toBlockHeader, BlockBody(Nil, Nil))
+    }
+
+    blockchain.save(genesisBlock)
+    blockchain.save(genesisBlock.header.hash, Nil)
+    blockchain.save(genesisBlock.header.hash, genesisBlock.header.difficulty)
+    genesisBlock
+  }
+
+  def loadInitialWorld(): InMemoryWorldStateProxy = {
+    InMemoryWorldStateProxy.persistState(getWorldState(scenario.pre))
+  }
+
+  def getFinalWorld(): InMemoryWorldStateProxy = {
+    InMemoryWorldStateProxy.persistState(getWorldState(scenario.postState))
+  }
+
+  def getBestBlock(): Option[Block] = {
+    val bestBlockNumber = storagesInstance.storages.appStateStorage.getBestBlockNumber()
+    blockchain.getBlockByNumber(bestBlockNumber)
+  }
+
+  private def buildBlockchainConfig(network: String): BlockchainConfig =  network match {
+    case "EIP150"     => new Eip150Config
+    case "Frontier"   => new FrontierConfig
+    case "Homestead"  => new HomesteadConfig
+    case "FrontierToHomesteadAt5" => new FrontierToHomesteadAt5
+    case "HomesteadToEIP150At5" => new HomesteadToEIP150At5
+
+    // Some default config, test will fail or be canceled
+    case _ => new FrontierConfig
+  }
+
+  private def decode(s: String): Array[Byte] = {
+    val stripped = s.replaceFirst("^0x", "")
+    Hex.decode(stripped)
+  }
+
+  // During decoding we cant expect some failures especially in bcInvalidRlPTests.json
+  private def decodeBlock(s: String): Option[Block] = {
+    Try(decode(s).toBlock) match {
+      case Success(block) => Some(block)
+      case Failure(ex) => {ex.printStackTrace(); None}
+    }
+  }
+
+  private def isInvalidBlock(blockDef: BlockDef): Boolean = {
+    blockDef.blockHeader.isEmpty && blockDef.transactions.isEmpty && blockDef.uncleHeaders.isEmpty
+  }
+
+  def getInvalid: List[BlockDef] = {
+    scenario.blocks.filter(isInvalidBlock)
+  }
+
+  def getBlocks(blocks: List[BlockDef]): List[Block] = {
+    blocks.flatMap(blockDef => decodeBlock(blockDef.rlp))
+  }
+
+  private def getWorldState(accounts: Map[Address, AccountState]): InMemoryWorldStateProxy = {
+    accounts.foldLeft(emptyWorld) { case (world, (address, accountState)) =>
+      val account = Account(nonce = accountState.nonce.u256, balance = accountState.balance.u256)
+      val worldWithAccountAndCode = world.saveAccount(address, account).saveCode(address, accountState.code)
+      val emptyStorage = worldWithAccountAndCode.getStorage(address)
+      val updatedStorage = accountState.storage.foldLeft(emptyStorage) { case (storage, (key, value)) =>
+        storage.store(key, value)
+      }
+      worldWithAccountAndCode.saveStorage(address, updatedStorage)
+    }
+  }
+
+  // TODO We need to take forks and branches into
+  // https://iohk.myjetbrains.com/youtrack/issue/EC-303
+  @tailrec
+  final def processBlocks(blocks: Seq[Block],
```

Comment:
Were my comments about this function addressed? Also what about the return type?

Reply:
Yes they were, function now only saves blocks in blockchain and current bestblock number. Both are needed, as we later need bestblock to check its state root.
What you mean by return type? It is `(Seq[NewBlock], Seq[BlockExecutionError])`

Reply:
👍 I must have looked at an old commit.

So why does it return `Seq[NewBlock]`? Where is it used?

Reply:
It is not necessary, actually only block errors can add some value.